controllers/helmrelease_controller_release.go

```diff
@@ -0,0 +1,426 @@
+	// most errors to be transient and return them.
+	values, err := r.composeValues(ctx, log, hr)
+	if err != nil {
+		v2.HelmReleaseNotReady(hr, v2.InitFailedReason, err.Error())
+		r.event(hr, hr.Status.LastAttemptedRevision, events.EventSeverityError, err.Error())
+		return ctrl.Result{}, err
+	}
+
+	// Observe the last release. If this fails, we likely encountered a
+	// transient error and should return it to requeue a reconciliation.
+	rls, err := run.ObserveLastRelease(hr)
+	if err != nil {
+		msg := "failed to determine last deployed Helm release revision"
+		v2.HelmReleaseNotReady(hr, v2.GetLastReleaseFailedReason, msg)
+		r.event(hr, hr.Status.LastAttemptedRevision, events.EventSeverityError, msg)
+		return ctrl.Result{}, err
+	}
+
+	// Determine the release strategy based on the existence of a
+	// release in the Helm storage.
+	makeRelease, remediation := run.Install, hr.Spec.GetInstall().GetRemediation()
+	successReason, failureReason := v2.InstallSucceededReason, v2.InstallFailedReason
+	if rls != nil {
+		makeRelease, remediation = run.Upgrade, hr.Spec.GetUpgrade().GetRemediation()
+		successReason, failureReason = v2.UpgradeSucceededReason, v2.UpgradeFailedReason
+	}
+
+	// Previous release attempt resulted in a locked pending state,
+	// attempt to remediate.
+	releaseRevision := util.ReleaseRevision(rls)
+	if rls != nil && hr.Status.LastReleaseRevision == releaseRevision && rls.Info.Status.IsPending() {
+		msg := fmt.Sprintf("previous release did not finish (%s)", rls.Info.Status)
+		v2.HelmReleaseNotReady(hr, v2.RemediatedCondition, msg)
+		r.event(hr, hr.Status.LastAttemptedRevision, events.EventSeverityError, msg)
+		remediation.IncrementFailureCount(hr)
+		return ctrl.Result{RequeueAfter: hr.Spec.Interval.Duration}, nil
+	}
+
+	valuesChecksum := util.ValuesChecksum(values)
+
+	// We must have gathered all information to decide if we actually
+	// need to move forward with the release at this point.
+	switch {
+	// Determine if there are any state changes to things we depend on
+	// (chart revision, values checksum), or if the revision of the
+	// release does not match what we have run ourselves.
+	case v2.StateChanged(hr, hc.GetArtifact().Revision, releaseRevision, valuesChecksum):
+		v2.HelmReleaseProgressing(hr)
+		r.patchStatus(ctx, hr)
+	// The state has not changed and the release is not in a failed state.
+	case rls != nil && remediation.GetFailureCount(*hr) == 0:
+		v2.HelmReleaseReady(hr, rls.Info.Description)
+		return ctrl.Result{RequeueAfter: hr.Spec.Interval.Duration}, nil
+	// We have exhausted our retries.
+	case remediation.RetriesExhausted(*hr):
+		v2.HelmReleaseNotReady(hr, meta.ReconciliationFailedReason, "exhausted release retries")
+		return ctrl.Result{RequeueAfter: hr.Spec.Interval.Duration}, nil
+	// Our previous reconciliation attempt failed, skip release to retry.
+	case hr.Status.LastSuccessfulReleaseRevision > 0 && hr.Status.LastReleaseRevision != hr.Status.LastSuccessfulReleaseRevision:
```

Comment:
The `LastSuccessfulReleaseRevision` seems to be managed by `reconcileTest`, why not check for the `Released==True` here instead, which is managed by this method (`reconcileRelease`)?

Reply:
While working on this, it became cumbersome to derive state from the conditions as a failure may or may not result in a new revision, and precise reflection of this is hard.

By opting to keep track of revision numbers instead in combination with observations of the Helm storage writes, we can re-observe based on what is recorded in the Helm storage for those revision numbers, instead of having the blindly trust our observations from the past.